File: tests/support/subscriber.ts
import { createServer, type IncomingMessage, type Server, type ServerResponse } from 'node:http';

/**
 * A webhook subscriber, the way a consumer's own endpoint behaves: it receives
 * the POST, records the raw bytes (needed to verify the signature) and can be
 * told to fail so the unit's retry behaviour is exercised for real rather than
 * simulated.
 */
export interface ReceivedWebhook {
  headers: Record<string, string>;
  rawBody: Buffer;
  receivedAt: number;
}

export interface SubscriberHandle {
  port: number;
  received: ReceivedWebhook[];
  /** Status to answer with; a function form can fail the first N deliveries. */
  respondWith: number | ((index: number) => number);
  close(): Promise<void>;
}

export async function startSubscriber(): Promise<SubscriberHandle> {
  const received: ReceivedWebhook[] = [];
  const handle: Partial<SubscriberHandle> & { received: ReceivedWebhook[]; respondWith: SubscriberHandle['respondWith'] } = {
    received,
    respondWith: 200,
  };

  const server: Server = createServer((req: IncomingMessage, res: ServerResponse) => {
    const chunks: Buffer[] = [];
    req.on('data', (c: Buffer) => chunks.push(c));
    req.on('end', () => {
      const headers: Record<string, string> = {};
      for (const [k, v] of Object.entries(req.headers)) {
        if (v !== undefined) headers[k.toLowerCase()] = Array.isArray(v) ? v.join(', ') : v;
      }
      const index = received.length;
      received.push({ headers, rawBody: Buffer.concat(chunks), receivedAt: Date.now() });
      const status = typeof handle.respondWith === 'function' ? handle.respondWith(index) : handle.respondWith;
      res.writeHead(status, { 'content-type': 'text/plain' });
      res.end(status >= 200 && status < 300 ? 'ok' : 'nope');
    });
  });

  await new Promise<void>((resolve) => server.listen(0, '0.0.0.0', () => resolve()));
  const address = server.address();
  const port = typeof address === 'object' && address ? address.port : 0;

  return Object.assign(handle, {
    port,
    close: () =>
      new Promise<void>((resolve) => {
        server.closeAllConnections?.();
        server.close(() => resolve());
      }),
  }) as SubscriberHandle;
}

File: tests/pytest/subscriber.py
"""A webhook subscriber that behaves like a consumer's own endpoint.

Records the RAW request bytes (the signature covers them, so a re-serialized
body would not verify) and can reject deliveries so the unit's retry behaviour
is exercised across a real socket.
"""

from __future__ import annotations

import threading
from dataclasses import dataclass, field
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from typing import Callable


@dataclass
class Received:
    headers: dict[str, str]
    raw_body: bytes


@dataclass
class Subscriber:
    port: int
    received: list[Received] = field(default_factory=list)
    respond_with: int | Callable[[int], int] = 200
    _server: ThreadingHTTPServer | None = field(default=None, repr=False)
    _thread: threading.Thread | None = field(default=None, repr=False)

    def status_for(self, index: int) -> int:
        return self.respond_with(index) if callable(self.respond_with) else self.respond_with

    def close(self) -> None:
        if self._server is not None:
            self._server.shutdown()
            self._server.server_close()
        if self._thread is not None:
            self._thread.join(timeout=5)


def start_subscriber() -> Subscriber:
    state = Subscriber(port=0)

    class Handler(BaseHTTPRequestHandler):
        protocol_version = "HTTP/1.1"

        def do_POST(self) -> None:  # noqa: N802 - required by BaseHTTPRequestHandler
            length = int(self.headers.get("content-length", "0"))
            raw = self.rfile.read(length)
            index = len(state.received)
            state.received.append(Received(headers={k.lower(): v for k, v in self.headers.items()}, raw_body=raw))
            status = state.status_for(index)
            payload = b"ok" if 200 <= status < 300 else b"nope"
            self.send_response(status)
            self.send_header("content-type", "text/plain")
            self.send_header("content-length", str(len(payload)))
            self.end_headers()
            self.wfile.write(payload)

        def log_message(self, *_args: object) -> None:
            """Silence the default stderr access log."""

    server = ThreadingHTTPServer(("0.0.0.0", 0), Handler)
    state.port = server.server_address[1]
    state._server = server
    thread = threading.Thread(target=server.serve_forever, daemon=True)
    thread.start()
    state._thread = thread
    return state
